patch_sm/Torus/torus.cpp

```diff
@@ -0,0 +1,412 @@
+
+void AudioCallback(AudioHandle::InputBuffer  in,
+                   AudioHandle::OutputBuffer out,
+                   size_t                    size)
+{
+    Settings&        settings_data = storage.GetSettings();
+    PerformanceState performance_state;
+    Patch            patch;
+
+    ProcessControls(&patch, &performance_state);
+    cv_scaler.Read(&patch, &performance_state, &calibration.cal);
+
+    if(settings_data.easter_egg_on)
+    {
+        for(size_t i = 0; i < size; ++i)
+        {
+            input[i] = in[0][i];
+        }
+        strummer.Process(NULL, size, &performance_state);
+        string_synth.Process(
+            performance_state, patch, input, output, aux, size);
+    }
+    else
+    {
+        // Apply noise gate.
+        for(size_t i = 0; i < size; i++)
+        {
+            float in_sample = in[0][i];
+            float error, gain;
+            error = in_sample * in_sample - in_level;
+            in_level += error * (error > 0.0f ? 0.1f : 0.0001f);
+            gain     = in_level <= kNoiseGateThreshold
+                           ? (1.0f / kNoiseGateThreshold) * in_level
+                           : 1.0f;
+            input[i] = gain * in_sample;
+        }
+
+        strummer.Process(input, size, &performance_state);
+        part.Process(performance_state, patch, input, output, aux, size);
+    }
+
+    for(size_t i = 0; i < size; i++)
+    {
+        out[0][i] = output[i];
+        out[1][i] = aux[i];
+    }
+}
+
+int main(void)
+{
+    hw.Init();
+    hw.SetAudioBlockSize(kMaxBlockSize);
+    float samplerate = hw.AudioSampleRate();
+    float blocksize  = hw.AudioBlockSize();
+    InitResources();
+
+    mode_button.Init(hw.B7,
+                     samplerate,
+                     Switch::TYPE_MOMENTARY,
+                     Switch::POLARITY_INVERTED,
+                     Switch::PULL_UP);
+    shift_switch.Init(hw.B8,
+                      samplerate,
+                      Switch::TYPE_TOGGLE,
+                      Switch::POLARITY_NORMAL,
+                      Switch::PULL_UP);
+
+    strummer.Init(0.01f, samplerate / blocksize);
+    part.Init(reverb_buffer);
+    string_synth.Init(reverb_buffer);
+
+    storage.Init(default_settings);
+    if(storage.GetState() == PersistentStorage<Settings>::State::FACTORY)
+    {
+        // Update the user values with the defaults.
+        storage.RestoreDefaults();
+    }
+    /** Restore settings from previous power cycle */
+    Settings& settings_data = storage.GetSettings();
+    calibration.cal.SetData(settings_data.scale, settings_data.offset);
+    polyListValue.SetIndex(settings_data.poly_mode);
+    eggListValue.SetIndex(settings_data.egg_mode);
+    modelListValue.SetIndex(settings_data.model_mode);
+    cv_scaler.Init();
+
+    hw.StartAdc();
+    hw.StartAudio(AudioCallback);
+
+    while(1)
+    {
+        // Without this pause the save events get missed for some unknown reason.
+        hw.Delay(1);
```

Comment:
I'd love to know what's going on here. If I don't have this delay and `trigger_save` becomes true it never gets picked up. I only discovered this because it worked if I had other processing happening in the while loop. I tried throttling with a timer but that had no effect.